### ISSUE TITLE
[6.1.1] False boolean values are being filtered out of values

### ISSUE THREAD

**Issue by clayne11** [erikras/redux-form#1993]
Any false value that gets passed as a payload gets filtered out and never gets set in the form `values` hash.

**Comment by kbzak** [comment]
I also have this problem - can't really set field value to false.

**Comment by kbzak** [comment]
For now my hacky workaround is to set booleans as 0 (false) and 1 (true) and then convert these when sending to API. But IMO redux-form should explicitly set fields to false, when asked to.

**Comment by bmv437** [comment]
This probably has something to do with a recent change in v6.1.0

https://github.com/erikras/redux-form/blob/master/src/reducer.js#L135
relevant issues #1774 #1762

I don't really agree with the fix, and not being able to set a value to false feels like a bug, but it was done on purpose.

**Comment by iyn** [comment]
I totally agree with what @clayne11 said. I explicitly need to store boolean value in the state and this is a breaking change.

**Comment by jole78** [comment]
I also agree. I have a custom component that needs to be able to set 3 different values actually [true, false, null/undefined] on a field and now it doesn't work.

**Comment by bmv437** [comment]
@jole78 I have the exact same issue.

@erikras can we get your thoughts?

**Comment by clayne11** [comment]
I'm going to create a PR to revert the changes. This type of logic should be left up to the application. Cutting values out of the library doesn't make sense and is very limiting.

**Comment by bbenezech** [comment]
@clayne11 @bmv437 Problem was:
- 2 states (undefined and false) that present the same UX (unchecked)
- inconsistency with text fields in RF (empty values aren't present in the values hash)
- divergence from the checkbox's HTML spec (no `False` state for checkboxes, they either submit their value when checked or don't submit anything)

That lead to cumbersome behavior with dirty/pristine/initialization management. (you had to manually set your checkbox fields to false at initialization to allow user to reverted them to pristine state).

This had me think that `False` is not an acceptable value for a checkbox field in RF, and that ReduxForm Checkboxes should toggle between `undefined <-> True`, not `False <-> True` nor `undefined -> True <-> False`.

The fact that RF allowed the latest 2 behaviors was implementation related and never intended as such (AFAIK).

@jole78 @kbzak UX-wise, you can replace the checkbox with a couple of radio buttons in a group. It will provide the old RF behavior with the correct UX (IMO).

I do agree it could have triggered a new minor version, I should have warned Erik about this.
Really sorry for everyone who relied on prior behavior and suffered a semver breach, I totally overlooked it.

**Comment by clayne11** [comment]
The issue is that the change you made doesn't just affect checkboxes - it affects every single form field. I use radio buttons with a `true` for yes and a `false` for no. This forces the user to make a selection. I don't want them to be able to leave the field undefined, as that would mean they missed the field. When a field is required, the user needs to specifically answer the question. This shows they read it and have actively put in a value for it.

`deleteInWithCleanup` deletes the value at a given path, and then if the parent object is empty is will delete the parent entry as well. You can't simply set anything to undefined as `redux-form` supports both ImmutableJS as well as regular JS objects so all operations on the `redux` store have been abstracted into helper methods. When you instantiate the form or reducer you need to import the correct type (either Immutable or plain JS) to initialize the form with the correct set of object modifier functions.

**Comment by bbenezech** [comment]
@clayne11 Yes. The same way you cannot store empty strings (unless it was initialized to something different from undefined). This is the issue at hand, I agree.

I understand `deleteInWithCleanup` should stay the way it is as it would be too complex to change.

The last 2 solutions left I see are: 
1. tweak dirty recursion so that `{someString: '', someBoolean: false}` is equivalent to `{}`.
2. consider '' and false to be different from undefined

Solution 1. requires deleting https://github.com/erikras/redux-form/pull/1774/files#diff-46f56fee267eb512b656d27b24987144R134 and https://github.com/erikras/redux-form/pull/1774/files#diff-46f56fee267eb512b656d27b24987144R152 and fixing dirty check recursion

Solution 2. requires deleting https://github.com/erikras/redux-form/pull/1774/files#diff-46f56fee267eb512b656d27b24987144R134 and https://github.com/erikras/redux-form/pull/1774/files#diff-46f56fee267eb512b656d27b24987144R152 but is a breaking change

**Comment by bbenezech** [comment]
Your use-case (radio group required with true/false) is the only valid use case in theory, but using boolean with radios is a weird idea. You really should use text in buttons' values as per the HTML spec.
Redux-Form uses boolean in value for checkable components (radios and checkboxes), but what it really means is checked when true, not checked otherwise, and only atomically for each individual checkable.

Anyway,
```javascript
     if (initial === undefined && (payload === '' || payload === false)) {
       result = deleteInWithCleanUp(result, `values.${field}`)
     else {...
```
was a really nasty hack.
I am for removing it and fixing dirty recursion for good.

**Comment by gilesbutler** [comment]
I've just found this issue after an hour trying to figure out why I couldn't parse my radio groups values to boolean. None of my team think using booleans with radios is a weird idea at all @bbenezech - so many forms have radio options of yes/no - it makes perfect sense to use booleans for those values. End users are familiar with using radio groups. It doesn't make any sense to use buttons when a radio group is the normal way of handling that input. Plus you then have to style the buttons for different states, selected vs not selected.

**Comment by clayne11** [comment]
Yeah, I think we'e agreed at this point that we need to switch back to the old behavior. Someone just has to find time to put together a PR to fix it.

**Comment by bbenezech** [comment]
@gilesbutler I don't get you. If you initialize your booleans with false, you can have a false/true toggling.
The issue is with ternary operators (undefined->True<->False).
@clayne11 RF needs to stop being a smart ass about which values it stores and deal correctly with all equivalences in the dirty recursion (including absent hash keys), which might not be a peace of cake.

**Comment by bmv437** [comment]
I have to agree with @clayne11 and @gilesbutler on this one, this was a big change.

In my case I have a radio field with three options: `Yes` `No` and `Unknown` where the underlying values are `true`, `false`, and (`undefined`/`null`). It used to work and now it's effectively broken because now `false === undefined` and you can't set a field to `false` unless it's initialValue was false (which may not always be the case).

I see what you're saying @bbenezech regarding pristine state of a form and checkboxes. If you don't specify initialValues for the checkbox, the state tree start off empty. Toggling a checkbox from unchecked -> checked set's the value in the state as `true`, then toggling it again from checked -> unchecked sets it to `false`. which means that the form is still dirty even though visually nothing changed. the state in the beginning was empty, and now it's `{ something: false }`

Still, I don't think that your solutions was the right one @bbenezech, and there should have been more discussion before the merge. having to manually set your checkbox initialValues to false isn't the worst thing in the world, or maybe you could use the `parse` prop for `Field` components to change `false` -> `undefined` when parsing a field value into the state tree. Changing core functionality of the library to fit your specific need isn't a fair solution.

@erikras It would be great get your perspective on this issue.

**Comment by bbenezech** [comment]
`Changing core functionality of the library to fit your specific need isn't a fair solution.`
I chose this implementation to keep consistency with pre-existing behavior (with text and empty values). I stated twice that this is not ideal but cannot be improved consistently until someone writes a proper dirty check that won't choke on non-existent keys/falsy values comparison.
I even wrote a first naive PR that had no effect because of this very issue.

The second PR (the one you all have an issue with) is a coherent and consistent addition over a non-ideal existing implementation: https://github.com/erikras/redux-form/pull/1774/files#diff-46f56fee267eb512b656d27b24987144R134

**Comment by erikras** [comment]
I agree with @clayne11. PR welcome.

**Comment by bbenezech** [comment]
@erikras You don't need a PR, you can just revert, right from github. https://github.com/erikras/redux-form/compare/erikras:master...bbenezech:revert-1774-master?expand=1

**Comment by erikras** [comment]
Duh... :facepalm:

**Comment by bbenezech** [comment]
@erikras Maybe we should remove `if (initial === undefined && payload === '') { `. It's a really dubious.

Depending on how you initialize the field, you get 2 different behaviors: 
- no init: `undefined -> 'a' <-> undefined`
- init to 'a': 'a' <-> ''

I don't know if you read the whole thread, but if you didn't, this would require to fix `dirty` recursion to return false for the comparison between having no key for a value, and a falsy value.
Unless we do that, dirty would be "broken" (emptying a text field would not restore it to pristine)

The problem at hands here would consistently be fixed, thanks to https://github.com/erikras/redux-form/pull/1762

**Comment by erikras** [comment]
> Unless we do that, dirty would be "broken" for text like it is now with booleans (emptying a text field would not restore it to pristine)

What if we taught `INITIALIZE` to treat `''` the same as `undefined` like `BLUR` does? That might be kind of complicated, though.

**Comment by clayne11** [comment]
I agree that we should stop initializing fields with `''`. It doesn't make sense unless the field's value is a string, which is only true some of the time. A lot of my fields are objects / dates / numbers / etc. I don't think we should be initializing fields to a default value at all.

**Comment by erikras** [comment]
[Here](https://github.com/erikras/redux-form/blob/49d0bc58f96dc355b397576f5838dcd2bcb63c3f/src/structure/plain/deepEqual.js#L5-L6) is where the dirty calculation is taking place. It _does_ seem to be equating `null`, `''`, and `false`.

**Comment by bmv437** [comment]
@erikras we also need to revert #1762

**Comment by erikras** [comment]
Revert published in [`v6.2.0`](https://github.com/erikras/redux-form/releases/tag/v6.2.0).

**Comment by GautierT** [comment]
Changing from `null -> false` doesn't trigger a rerender but it's changed in the store. 
Changing from `false -> undefined/empty` does nothing.
Changing from `false -> null` doesn't trigger a rerender but it's changed in the store. 
Is this normal behavior ?
How can i set a field to `false` and have it display things based on this ?
related to #2600 
Thanks

**Comment by GautierT** [comment]
Hi all,
Sorry for the spam but i don't understand how to have a rerender after setting a field to `false`. 
I can see the change in the store (`null` to `false`) but my component don't rerender. it rerender only when i go `true` to `false`.
Help please ! 😳

**Comment by jacklj** [comment]
This was never reverted, was it? Seems like the described behaviour still occurs. Thanks!

**Comment by danielmackey** [comment]
@jacklj it looks like regession in 6.7.0. We locked out version to 6.6.3 to work around. I commented here in the newer issue with examples on how to verify the regression. https://github.com/erikras/redux-form/issues/2922

**Comment by cescoferraro** [comment]
@danielmackey that solved!  6.8.0 still contains the regression

**Comment by blord-fullscreen** [comment]
I am still seeing this behavior in 6.8.0. I initialize a checkbox with `false`, and the value is converted to an empty string in the values hash. Double-toggling back and forth in the UI and watching in redux dev tools shows the same thing: the checkbox only has `true` and empty string as possible values. This is really maddening, for all the reasons described above. Give me back `false`!

I am using other features in 6.8.0 so can't downgrade easily. (I may just write a little utility to walk the values tree and coerce all keys that begin with 'is' to Boolean. Yuck.)

**Comment by bmv437** [comment]
@blord-fullscreen try this
https://github.com/erikras/redux-form/issues/2922#issuecomment-311935389
`<Field component="input" name="currently" id="currently" type="checkbox" normalize={v => !!v} />`
the normalize function can force false when empty

**Comment by erikras** [comment]
⚠️ __IMPORTANT__ ⚠️  As of [`v7.0.0`](https://github.com/erikras/redux-form/releases/tag/v7.0.0), `undefined` and `false` are no longer equal. [Read more](https://github.com/erikras/redux-form/releases/tag/v7.0.0).

TLDR: @clayne11 was right all along. 😆
